### PR TITLE
Gcc 9.3.1 update

### PR DESCRIPTION
**Pull request by quzard**

### CODE REVIEW

File: core/CMakeLists.txt [hunk 1]
```diff
@@ -17,7 +17,7 @@ project(logtail)
 
 # Options.
 option(BUILD_LOGTAIL_UT "Build unit test for Logtail" OFF)
-option(ENABLE_COMPATIBLE_MODE "Build Logtail in compatible mode (for low version Linux)" OFF)
+option(ENABLE_COMPATIBLE_MODE "Build Logtail in compatible mode (for low version Linux)" ON)
```
Comment (yyuuttaaoo) [inline core/CMakeLists.txt:20]: don't modify this

File: core/deps/spdlog/sinks/rotating_file_sink-inl.h
Comment (yyuuttaaoo) [inline core/deps/spdlog/sinks/rotating_file_sink-inl.h:1]: 内置的rotate on open不能用吗？能用就不用这个了

File: .devcontainer/Dockerfile [hunk 2]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-FROM sls-opensource-registry.cn-shanghai.cr.aliyuncs.com/ilogtail-community-edition/ilogtail-build-linux:latest
+FROM sls-opensource-registry.cn-shanghai.cr.aliyuncs.com/ilogtail-community-edition/ilogtail-build-linux:gcc_9.3.1_test
```
Comment (yyuuttaaoo) [inline .devcontainer/Dockerfile:15]: 需要一个多架构的去掉_test的镜像

File: core/observer/network/protocols/common.h [hunk 3]
```diff
@@ -26,6 +26,17 @@
 #include <unordered_map>
 #include <ostream>
 
+static void log_trace_wrapper(int64_t mHeadRequestsIdx, int64_t mTailRequestsIdx, int64_t mTailResponsesIdx) 
+{
+    do {
+        if (true && sLogger->should_log(spdlog::level::trace)) {
```
Comment (yyuuttaaoo) [inline core/observer/network/protocols/common.h:32]: Will the original code compile under 9.3.1?

File: docker/Dockerfile.ilogtail-toolchain-linux [hunk 4]
```diff
@@ -65,6 +65,25 @@ RUN yum -y install gcc gcc-c++ make libuuid-devel libstdc++-static systemd-devel
...
+    PERL5LIB=/opt/rh/devtoolset-9/root//usr/lib64/perl5/vendor_perl:/opt/rh/devtoolset-9/root/usr/lib/perl5:/opt/rh/devtoolset-9/root//usr/share/perl5/vendor_perl \
+    X_SCLS=devtoolset-9 \
+    PCP_DIR=/opt/rh/devtoolset-9/root \
+    LD_LIBRARY_PATH=/opt/rh/devtoolset-9/root/usr/lib64:/opt/rh/devtoolset-9/root/usr/lib:/opt/rh/devtoolset-9/root/usr/lib64/dyninst:/opt/rh/devtoolset-9/root/usr/lib/dyninst:/opt/rh/devtoolset-9/root/usr/lib64:/opt/rh/devtoolset-9/root/usr/lib \
+    PATH=/usr/ali/bin:/opt/rh/devtoolset-9/root/usr/bin:$PATH \
+    PYTHONPATH=/opt/rh/devtoolset-9/root/usr/lib64/python2.7/site-packages:/opt/rh/devtoolset-9/root/usr/lib/python2.7/site-packages \
+    PKG_CONFIG_PATH=/opt/rh/devtoolset-9/root/usr/lib64/pkgconfig \
+    INFOPATH=/opt/rh/devtoolset-9/root/usr/share/info
+
+RUN yum makecache
```
Comment (yyuuttaaoo) [inline docker/Dockerfile.ilogtail-toolchain-linux:85]: is line 85 and 86 essential?

File: docker/Dockerfile.ilogtail-toolchain-linux [hunk 5]
```diff
@@ -65,6 +65,25 @@ RUN yum -y install gcc gcc-c++ make libuuid-devel libstdc++-static systemd-devel
 RUN [[ ${TARGETPLATFORM##*/} == 'amd64' ]] && ARCH='x86_64' || ARCH='aarch64' && \
     debuginfo-install -y glibc-2.17-326.el7_9.${ARCH} libuuid-2.23.2-65.el7_9.1.${ARCH}
 
+# install dependencies
+RUN yum -y install epel-release centos-release-scl
+RUN yum -y install devtoolset-9-gcc devtoolset-9-gcc-c++ make
+RUN yum -y install curl-devel expat-devel gettext-devel openssl-devel perl-devel python3-devel zlib-devel \
```
Comment (yyuuttaaoo) [inline docker/Dockerfile.ilogtail-toolchain-linux:71]: is this required?

File: docker/Dockerfile.ilogtail-toolchain-linux [hunk 6]
```diff
@@ -65,6 +65,25 @@ RUN yum -y install gcc gcc-c++ make libuuid-devel libstdc++-static systemd-devel
 RUN [[ ${TARGETPLATFORM##*/} == 'amd64' ]] && ARCH='x86_64' || ARCH='aarch64' && \
     debuginfo-install -y glibc-2.17-326.el7_9.${ARCH} libuuid-2.23.2-65.el7_9.1.${ARCH}
 
+# install dependencies
+RUN yum -y install epel-release centos-release-scl
```
Comment (yyuuttaaoo) [inline docker/Dockerfile.ilogtail-toolchain-linux:69]: please merge line 69 and 70 into one and append the following code to reduce the image size.
``` shell
&& yum -y clean all && rm -fr /var/cache && rm -rf /core.*
```

File: scripts/gen_build_scripts.sh [hunk 7]
```diff
@@ -33,7 +33,7 @@ PLUGINS_CONFIG_FILE=${8:-${PLUGINS_CONFIG_FILE:-plugins.yml,external_plugins.yml
 GO_MOD_FILE=${9:-${GO_MOD_FILE:-go.mod}}
 
 BUILD_LOGTAIL_UT=${BUILD_LOGTAIL_UT:-OFF}
-ENABLE_COMPATIBLE_MODE=${ENABLE_COMPATIBLE_MODE:-OFF}
```
Comment (yyuuttaaoo) [inline scripts/gen_build_scripts.sh:36]: not required. the switch will be turn on in git workflow

**Review by yyuuttaaoo — CHANGES_REQUESTED** [review]

File: .github/workflows/build-core.yaml [hunk 8]
```diff
@@ -32,7 +32,7 @@ on:
 jobs:
   CI:
     runs-on: ${{ matrix.runner }}
-    timeout-minutes: 60
+    timeout-minutes: 100
```
[the remark this answers is not in the document]
Comment (quzard) [reply .github/workflows/build-core.yaml:35]: 一小时够了，我只是测试时怕超时，就都改高了。

File: .github/workflows/build-core.yaml [hunk 9]
```diff
@@ -44,7 +44,19 @@ jobs:
       - name: prepare ubuntu environment
         if: matrix.runner == 'ubuntu-latest'
         run: sudo apt-get clean && sudo apt-get update && sudo apt-get install -y libsystemd-dev
-
+      - name: clean
```
[the remark this answers is not in the document]
Comment (quzard) [reply .github/workflows/build-core.yaml:47]: action会有几率报容量不够的错误。

File: core/CMakeLists.txt (hunk 1 above)
Comment (quzard) [reply core/CMakeLists.txt:20]: ok

File: core/deps/spdlog/sinks/rotating_file_sink-inl.h
Comment (quzard) [reply core/deps/spdlog/sinks/rotating_file_sink-inl.h:1]: 之前测试了一下，内置的rotate on open不能用。

File: .devcontainer/Dockerfile (hunk 2 above)
Comment (quzard) [reply .devcontainer/Dockerfile:15]: ok

File: core/observer/network/protocols/common.h (hunk 3 above)
Comment (quzard) [reply core/observer/network/protocols/common.h:32]: 不修改编译会不通过

Comment (quzard) [reply core/observer/network/protocols/common.h:32]: /workspaces/ilogtail_github/core/observer/network/protocols/common.h: In member function 'bool logtail::CommonCache<reqType, respType, aggregatorType, eventType, capacity>::TryStitcherByResp()':
/workspaces/ilogtail_github/core/observer/network/protocols/common.h:384:94: error: expression cannot be used as a function
  384 |                       ("head_req", this->mHeadRequestsIdx)("tail_req", this->mTailRequestsIdx)(
      |                                                                                              ^
/workspaces/ilogtail_github/core/observer/network/protocols/common.h:383:13: error: there are no arguments to 'LOG_TRACE' that depend on a template parameter, so a declaration of 'LOG_TRACE' must be available [-fpermissive]
  383 |             LOG_TRACE(sLogger,
      |             ^~~~~~~~~
/workspaces/ilogtail_github/core/observer/network/protocols/common.h:383:13: note: (if you use '-fpermissive', G++ will accept your code, but allowing the use of an undeclared name is deprecated)

**Review by messixukejia — APPROVED** [review]

**Review by yyuuttaaoo — APPROVED** [review]

**Pull request merged**